### PR TITLE
Reduce RAM usage during session renegotiation

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR reduces RAM consumption during session renegotiation by not storing two instances of the peer's CRT chain and session ticket at the same time. This is based on #2411.

__Gatekeeper note:___ (added by mpg): please merge #2411 before this one, but the two can be merged in the same round.

__Details:__ During a handshake, Mbed TLS constructs an instance of   `mbedtls_ssl_session` representing the session being established. This structure contains information such as the session's master secret, the peer certificate, or the session ticket issues by the server (if applicable). During a renegotiation, the instance representing the new session is constructed alongside the existing one and replaces the latter only when the renegotiation is complete. While conceptually clear, this leads to large pieces of information such as the peer's CRT or the session ticket being present twice in memory at the same time, even though the original instances are removed eventually (and are, in the case of the peer CRT, identical). This PR improves on that by avoiding the simultaneous storage of two CRT chains or session tickets in RAM.

For the peer CRT, care has to be taken (a) not to change observable behaviour, and  (b) to keep the mitigation against the [triple handshake attack](https://www.mitls.org/pages/attacks/3SHAKE). Please see the commit messages for more detail. 

The amount of RAM saved is the sum of the size of the session ticket and the size of the peer's CRT chain, which may vary widely with the application context. Using default tickets and test certificates from Mbed TLS, for example, the savings amount to around 2Kb.

### CODE REVIEW

**Comment by mpg** [comment]
CI results: travis passed, and Jenkins failed only in known-fragile timing test on BSD, so as good as a pass.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -5435,6 +5435,23 @@ int mbedtls_ssl_write_certificate( mbedtls_ssl_context *ssl )
...
+{
+    mbedtls_x509_crt const * const peer_crt = ssl->session->peer_cert;
+
+    if( peer_crt == NULL )
+        return( -1 );
+
+    if( peer_crt->raw.len != crt_buf_len )
+        return( -1 );
+
+    return( memcmp( peer_crt->raw.p, crt_buf, crt_buf_len) );
```
Comment (k-stachowiak) [inline library/ssl_tls.c]: There should be a space between `crt_buf_len` and `)`.

File: library/ssl_tls.c [hunk 2]
```diff
@@ -5525,116 +5542,142 @@ static int ssl_parse_certificate_chain( mbedtls_ssl_context *ssl )
...
-        switch( ret )
+        /* Check if we're handling the first CRT in the chain. */
+        if( ssl->session_negotiate->peer_cert == NULL )
         {
-        case 0: /*ok*/
-        case MBEDTLS_ERR_X509_UNKNOWN_SIG_ALG + MBEDTLS_ERR_OID_NOT_FOUND:
-            /* Ignore certificate with an unknown algorithm: maybe a
-               prior certificate was already trusted. */
-            break;
+            /* During client-side renegotiation, check the server's end-CRTs
```
Comment (k-stachowiak) [inline library/ssl_tls.c]: Maybe this sentence should say something like "check if" or "check that", rather than plain "check"?

**Comment by hanno-arm** [comment]
Ping @sbutcher-arm and @Patater for gatekeeping.

File: library/ssl_tls.c (hunk 1 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Thanks, I'll fix it.

File: library/ssl_tls.c (hunk 2 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Yes, I'll change it.

**Comment by hanno-arm** [comment]
@k-stachowiak I think I have addressed your comments. Could you please re-review?

**Comment by hanno-arm** [comment]
@mpg @k-stachowiak I rebased this on top of #2391. While there is no dependency this PR and #2391, later optimization PRs will need both this and #2391, and with merge commits being strongly discouraged, creating a linear history between the PRs seems to be the only way to go forward.

**Comment by mpg** [comment]
@hanno-arm There must be some kind of confusion. This is 2391, and you're stating that you rebased it on top of 2391. I'm not sure if you mean to rebase another PR on top of this one, or this one on top of another. Either way, please clarify.

**Comment by hanno-arm** [comment]
@mpg I meant #2411, sorry!

**Comment by mpg** [comment]
Travis passed and Jenkins only failed in the know-fragile timing test on BSD, so that's as good as a pass. The PR also has a ChangeLog entry and pre-existing tests, so it nearly "ready for merge" except for the dependency on #2411.

**Pull request merged**